### PR TITLE
feat: k2pow service

### LINKED ISSUE #343: Feature request: remote compute server (proving k2pow)
Request to allow offloading of k2pow to a separate server.
This problem was partially solved by the 1:N post-service feature, however it still relies upon compute being executed where storage exists or relying on suboptimal network share performance.

The idea of this feature is to allow separation of concerns between a low power storage server such as an Intel N100 and a high power compute node, such as a gaming computer or dedicated ex-enterprise server. 
This would provide security to the network, but lower electricity costs for smeshers.

As background information, this feature was added to h9-miner earlier in the year.
I would like to see the feature introduced for official software to reduce incentive to move to h9-miner and help support "free range" netspace.

### PR DESCRIPTION
**Pull request by acud**
Work in progress.
As discussed with NJ, this implementation will:
- allow users to run a k2pow compute worker as a separate binary
- the worker will have a very simple HTTP API
  - the API will have create, read, update operations for jobs
  - right now all is done in-memory for the worker, no persistence of state
- the post service needs a new flag that will start it in a special mode, in which jobs of k2pow computations will call a specified HTTP hook with the necessary params
  - how the result comes back to the post service is still TBD (either a hook on the service, or wait for response)

closes #343

### CODE REVIEW

File: src/prove.rs [hunk 1]
```diff
@@ -147,21 +147,28 @@ impl Prover8_56 {
...
 
+        let ciphers: Vec<AesCipher> = match pow_prover.par() {
+            true => nonce_group_range(nonces.clone(), Self::NONCES_PER_AES)
+                .into_par_iter()
+                .map(map_fn)
+                .collect::<eyre::Result<_>>()?,
+            false => nonce_group_range(nonces.clone(), Self::NONCES_PER_AES)
+                .map(map_fn)
+                .collect::<eyre::Result<_>>()?,
+        };
```
Comment (poszu) [inline src/prove.rs:171]: Now that I think of it again, using rayon has a downside of limiting parallel execution to the parallelism available on the **local** machine, while it should be limited by the number of available backend services. I.e. if the local machine has 2 cores it will run 2 k2pows in parallel, even if there are more k2pow services available. On the other hand, we can't know how many k2pow services there are. Perhaps it should just spawn a thread for every nonce_group that needs a k2pow? IDK if rayon supports "unlimited" parallelism.

File: src/prove.rs [hunk 2]
```diff
@@ -147,21 +147,28 @@ impl Prover8_56 {
...
-            .collect::<eyre::Result<_>>()?;
+        let map_fn = |nonce_group: u32| -> eyre::Result<AesCipher> {
+            log::debug!("calculating proof of work for nonce group {nonce_group}");
+            let pow = pow_prover.prove(
+                nonce_group.try_into()?,
+                challenge[..8].try_into().unwrap(),
+                &params.pow_difficulty,
+                miner_id,
+            )?;
+            log::debug!("proof of work: {pow}");
```
[the remark this answers is not in the document]
Comment (acud) [reply src/prove.rs]: was thinking that ideally those would be span fields but yeah this is fine too

File: k2pow-service/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+                    info_span!(
+                        "http_request",
+                        method = ?request.method(),
+                        uri =  matched_path,
+                        status = tracing::field::Empty,
+                    )
+                })
+                .on_response(|response: &Response, _latency: Duration, span: &Span| {
+                    span.record("status", response.status().as_str());
+                    tracing::info!("served request");
```
[the remark this answers is not in the document]
Comment (acud) [reply k2pow-service/src/main.rs:124]: So I actually like this in `info` because you can see that the server is doing something. It tells you it serves requests and so you know it's still alive and generating output/serving requests. It doesn't show any debugging information, just that the request was served and the status code. I like to think of it as something similar to nginx access log

File: k2pow-service/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+        randomx_flags |= post::pow::randomx::RandomXFlag::FLAG_LARGE_PAGES;
+    }
+    eprintln!("RandomX flags: {}", randomx_flags);
+
+    loop {
+        sleep(Duration::from_secs(POLL_EVERY_SECS)).await;
+        let job = match job_manager.take().await {
+            None => continue,
+            Some(v) => v,
+        };
```
[the remark this answers is not in the document]
Comment (acud) [reply k2pow-service/src/main.rs:165]: Yeah so I don't mind doing it this way, but I'd chuck away and throw the whole job manager with it if I'd do so because it doesn't make much sense in this case to have it. If the job manager already returns a handle then it's already coupled enough to just have the axum handler to take a `HashMap` as `State`, and just throw everything else to the dustbin.

> You would avoid having another thread just for starting the tasks

It's not a separate thread, but just an async task so I'm not sure the overhead is so big.

File: src/prove.rs (hunk 1 above)
Comment (acud) [reply src/prove.rs:171]: Yeah I had some doubts about how this would work. The problem is a bit more nuanced, and I think that `rayon` and the parallel iterator are misused here, since they aim to solve a different problem here.
The thing is that we don't need a thread pool here. Work stealing with little threads would do the same thing with far less context switching. `async` would be superior here, because:
- the request to the API that provides the response involves almost entirely of I/O on the network side.
- having multiple blocking threads on this sort of task is extremely wasteful.
- it provides back-pressure indeed, but in a very ugly way and with a huge overhead of the thread-pool.

Speaking purely on what we probably want in this specific case is the following:
- have a regular iterator that spawns multiple async tasks that call the k2pow service, but not `await` on them.
- no parallel iteration is needed in this case, it can go through them sequentially
- depending on how many of those requests we need to do, we could just spawn all the futures all at once, and the k2pow service will make sure to execute (, throttle) them and return the results. on the post-service side, we could just use something like [this](https://docs.rs/futures/latest/futures/future/fn.try_join_all.html) to get the results, ordered.
- the `cores` param turns into, instead, a parameter that would be throttling the number of open requests to remote k2pow service/LB. we could use a `tokio::sync::Semaphore` to do that. The k2pow client (in `src/pow/service.rs`) would in turn use this parameter, and will therefore limit the amount of open in-flight requests (it would be rather a `parallelism` setting for api calls).
- then zip the iterator of the (ordered) results with the iterator of the inputs, generating the same vector of `AesCipher` we'd normally get from the sync execution context when performing the calculation locally.

Now... speaking of weaving this into what we already have, I suppose is a bit more complex. It also means plumbing in `async` way deeper into the call-stack into areas where either synchronous-blocking execution may also happen and I'm not sure how I feel about it, because right now the async context is completely missing from those areas in the codebase which, I find, makes it easier to reason about it better.

Maybe just having a `parallelism` setting that will just spawn a constant size rayon threadpool on which to execute all the api calls on would be easier, but at a cost of a lot of unnecessary overhead of thread context-switching.

File: src/prove.rs [hunk 5]
```diff
@@ -130,11 +130,11 @@ pub struct Prover8_56 {
 impl Prover8_56 {
     pub(crate) const NONCES_PER_AES: u32 = 16;
 
-    pub fn new<P: pow::Prover>(
+    pub fn new(
         challenge: &[u8; 32],
         nonces: Range<u32>,
         params: ProvingParams,
-        pow_prover: &P,
+        pow_prover: Arc<dyn pow::Prover>,
```
[the remark this answers is not in the document]
Comment (acud) [reply src/prove.rs:137]: the problem is in `service/src/service.rs`, I can't really figure out how to do this without boxing in the value:
```rust
                let threads = self.threads.clone();
                let stop = self.stop.clone();
                let progress = ProvingProgress::default();
                let pow_prover: dyn post::pow::Prover = match &self.remote_k2pow {
                    Some(uri) => post::pow::service::K2powService::new(uri.clone()), <- breaks compilation
                    None => post::pow::randomx::PoW::new(pow_flags).unwrap(),
                };
                let reporter = progress.clone();
                *proof_gen = ProofGenProcess::Running {
                    challenge,
                    handle: Some(std::thread::spawn(move || {
                        post::prove::generate_proof(
                            &datadir,
                            &challenge,
                            cfg,
                            nonces,
                            threads,
                            pow_flags,
                            stop,
                            reporter,
                            &pow_prover,
                        )
                    })),
                    progress,
                };

```

File: k2pow-service/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+        .await
+        .unwrap();
+    axum::serve(listener, router).await.unwrap();
+
+    Ok(())
+}
+
+fn router<T: GetOrCreate>(job_manager: Arc<T>) -> Router {
+    Router::new()
+        .route("/", get(root))
```
[the remark this answers is not in the document]
Comment (acud) [reply k2pow-service/src/main.rs:104]: k8s health check endpoint?

File: src/prove.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (poszu) [reply src/prove.rs:137]: It needs to be boxed there because the two branches of the `match` return a different type. This is perfectly fine IMHO:
```rust
let pow_prover: Box<dyn post::pow::Prover> = match &self.remote_k2pow {
    Some(uri) => Box::new(post::pow::service::K2powService::new(uri.clone())),
    None => Box::new(post::pow::randomx::PoW::new(pow_flags).unwrap()),
};
```

File: k2pow-service/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (poszu) [reply k2pow-service/src/main.rs:124]: Logs that tell that something happened and are emitted on every request are usually trace-level. They have no value for the operator except for debugging.

File: k2pow-service/src/main.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (poszu) [reply k2pow-service/src/main.rs:165]: > If the job manager already returns a handle then it's already coupled enough to just have the axum handler to take a HashMap as State, and just throw everything else to the dustbin.

It doesn't need to return the handle. It might have an enum for its inner logic (keeping job's state), which would have the additional data like the job handle and a different one returned to the caller (this one would only say "in progress" vs "done, here's the result (or error)". Translation from the former to the latter is trivial.

> > You would avoid having another thread just for starting the tasks
>
> It's not a separate thread, but just an async task so I'm not sure the overhead is so big.

It's not about the performance but the unnecessary complexity of the whole thing. Btw, IMHO async is worse because the code gets more complicated quickly. It would be great if JobManager could be "sync".

File: src/prove.rs (hunk 1 above)
Comment (poszu) [reply src/prove.rs:171]: >The thing is that we don't need a thread pool here. Work stealing with little threads would do the same thing with far less context switching.

Rayon [uses work stealing](https://github.com/rayon-rs/rayon/blob/main/FAQ.md#how-does-rayon-balance-work-between-threads) :)

Regarding async, I don't feel this is a good idea to make `generate_proof` async. But maybe it would make sense to add another _blocking_ method to the `pow::Prover` trait that would calculate proofs of work for a given range of input nonces? Inside it could asynchronously spawn requests and await them. Btw, this is how blocking requests in `reqwest` work too (they expose a _sync_ API for the _async_ inner functions). See also a helpful post here: https://stackoverflow.com/questions/51044467/how-can-i-perform-parallel-asynchronous-http-get-requests-with-reqwest

>the cores param turns into, instead, a parameter that would be throttling the number of open requests to remote k2pow service.

I'd rather not repurpose existing configuration parameter doing other things for this. I think it would be best to have another configuration option for this with some sane default.

Comment (acud) [reply src/prove.rs:171]: > Rayon [uses work stealing](https://github.com/rayon-rs/rayon/blob/main/FAQ.md#how-does-rayon-balance-work-between-threads) :)

It does, in order to do a lot of _CPU-bound_ work (my understanding, also from reading around about rayon, is that it is meant for CPU bound use-cases, not IO bound). When we ask for a remote k2pow service to produce the results we need, there's zero computation involved, only waiting on the response of the http response. So, if you have, for example, 200 remote k2pow hosts to query from, you'll need 200 threads to do that in parallel, while in async context you'll need maybe a couple. That sounds pretty wasteful.

> Regarding async......

So here I'm not sure I'm fully onboard and this smells a bit, because:
- the main application is spawned as a tokio async application
- the tonic part is async as well
- the actual worker part is currently spawned as a blocking task within tokio (which is fine because currently it is 100% CPU bound)

The changes you're suggesting would mean (in the case of the k2pow service case):
- the main application is spawned as a tokio async application
- the tonic part is async as well
- worker part is currently spawned as a blocking task within tokio
- within the blocking task, a large threadpool is created
- within each thread, a blocking task is created
- the blocking task uses a blocking API that uses async under-the-hood

So to sum it up, I'm not 100% fan of spawning an async context in order to call a blocking context that calls an async context.

> I'd rather not repurpose existing configuration parameter doing other things for this. I think it would be best to have another configuration option for this with some sane default.

I meant that, sorry, my phrasing was somewhat misleading. I meant to use a differently named parameter.

File: src/prove.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (acud) [reply src/prove.rs:137]: Sure, but then how can `pow_prover` type be `pow_prover: &dyn pow::Prover` as you requested to change it? Shouldn't it be `pow_prover: Box<dyn pow::Prover>`?

File: k2pow-service/src/main.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (acud) [reply k2pow-service/src/main.rs]: we can do it this way yeah. I don't mind doing the change. All in all, I'm not sure how much code it will really save on EOD... But it's fine to do it this way :+1:

File: src/prove.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (poszu) [reply src/prove.rs]: You can just take a reference `&pow_prover`. It will work because `Box` [implements](https://doc.rust-lang.org/std/boxed/struct.Box.html#impl-Deref-for-Box%3CT,+A%3E) [`std::ops::Deref`](https://doc.rust-lang.org/std/ops/trait.Deref.html#) and because deref coercion. Specifically:

> If T implements Deref<Target = U>, and v is a value of type T, then:
> - Values of type &T are coerced to values of type &U

File: src/prove.rs (hunk 1 above)
Comment (poszu) [reply src/prove.rs:171]: I think we agree that rayon is not the best match here :)

About
>The changes you're suggesting would mean (in the case of the k2pow service case):
> - the main application is spawned as a tokio async application
> - the tonic part is async as well
> - worker part is currently spawned as a blocking task within tokio
> - within the blocking task, a large threadpool is created
> - within each thread, a blocking task is created
> - the blocking task uses a blocking API that uses async under-the-hood

I think you misunderstood me. I meant something like this:
...
- worker part is currently spawned as a blocking task within tokio
- within the blocking task (`generate_proof`), call blocking pow::Prover::ProveMany, spawn async task per nonce_group and join them
- the blocking task uses a blocking API that uses async under-the-hood (using something like https://docs.rs/futures/latest/futures/executor/fn.block_on.html to run the above future to completion on the current thread).

There is no additional thread pool involved. By the way, this also happens under the hood in the `reqwest` blocking API anyways.

File: k2pow-service/src/job_manager.rs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    pub difficulty: [u8; 32],
+    pub miner: [u8; 32],
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{GetOrCreate, JobError, JobState};
+
+    #[tokio::test]
+    async fn test_job_manager() {
```
Comment (poszu) [inline k2pow-service/src/job_manager.rs:93]: I think the test can be _sync_ now that `JobManager` is fully sync.
```suggestion
    #[test]
    fn test_job_manager() {
```

File: service/src/service.rs [hunk 8]
```diff
@@ -110,6 +110,9 @@ pub struct PostService {
     threads: post::config::Cores,
     pow_flags: RandomXFlag,
     proof_generation: Mutex<ProofGenProcess>,
+    remote_k2pow: Option<String>,
+    remote_k2pow_parallelism: usize,
+    remote_k2pow_backoff: u64,
```
Comment (poszu) [inline service/src/service.rs:115]: How about enclosing it in a Config struct?

```rust
pub struct K2powConfig {
    url: String,
    parallelilsm: usize
    backoff: time::Duration,
}
```

```suggestion
    remote_k2pow: Option<K2powConfig>,
```

File: service/src/service.rs [hunk 9]
```diff
@@ -122,6 +125,9 @@ impl PostService {
         nonces: usize,
         threads: post::config::Cores,
         pow_flags: RandomXFlag,
+        remote_k2pow: Option<String>,
+        remote_k2pow_parallelism: usize,
+        remote_k2pow_backoff: u64,
```
Comment (poszu) [inline service/src/service.rs:130]: Then it becomes
```suggestion
        remote_k2pow: Option<K2powConfig>,
```

File: service/tests/test_client.rs [hunk 10]
```diff
@@ -280,6 +280,9 @@ async fn test_get_metadata(#[case] vrf_difficulty: Option<[u8; 32]>) {
         16,
         post::config::Cores::Any(1),
         post::pow::randomx::RandomXFlag::get_recommended_flags(),
+        None,
+        1,
+        1,
```
Comment (poszu) [inline service/tests/test_client.rs:285]: And you wouldn't need to provide useless values for the case when remote is disabled:
```suggestion
        None,
```

File: Cargo.toml [hunk 11]
```diff
@@ -45,6 +46,9 @@ thread_local = "1.1.7"
 mockall = "0.13.0"
 core_affinity = "0.8.1"
 hex = "0.4.3"
+reqwest = { version = "0.12.7", features = ["blocking"] }
```
Comment (poszu) [inline Cargo.toml:49]: The blocking feature isn't needed anymore, is it?
```suggestion
reqwest = { version = "0.12.7"}
```

File: k2pow-service/src/job_manager.rs [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+            randomx_large_pages,
+        }
+    }
+    fn check_finished(&self, job: &Job) {
+        // there's probably a better way to write this with interior mutability on the
+        // value that the HashMap stores (like with a RefCell). For now, because the
+        // joining the handle consumes the value, borrowing doesn't work here and we need
+        // to remove the value, use it, then put it back into the map.
+        let mut hs = self.jobs.lock().unwrap();
+        let entry = hs.remove_entry(job);
```
Comment (poszu) [inline k2pow-service/src/job_manager.rs:61]: AFAIR, I used an `Option<JoinHandle>` for this reason before so I could `take()` handle out of it and later put it back.

File: k2pow-service/src/job_manager.rs [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+        self.check_finished(&job);
+        let mut hs = self.jobs.lock().unwrap();
+        match hs.get(&job) {
+            Some(JobState::InProgress(_)) => Ok(JobStatus::InProgress),
+            Some(JobState::Done(result)) => Ok(JobStatus::Done(result.clone())),
+            None => {
+                let other = hs.values().find(|v| matches!(v, JobState::InProgress(_)));
+                if other.is_some() {
+                    return Err(JobError::TooManyJobs);
+                }
```
Comment (poszu) [inline k2pow-service/src/job_manager.rs:96]: Perhaps it could hold also the `Job` that is currently processed to avoid linear search here? I mean something like:

```rust
struct Jobs {
  in_progress: Option<Job>,
  states: HashMap<Job, JobState>,
}

pub struct JobManager {
    jobs: Mutex<Jobs>,
```

Then this becomes:
```suggestion
                if hs.in_progress.is_some() {
                    return Err(JobError::TooManyJobs);
                }
```

File: src/pow/service.rs [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+pub struct K2powService {
+    k2pow_service: String,
+    semaphore: Arc<Semaphore>,
+    backoff: Duration,
+}
+
+impl K2powService {
+    pub fn new(k2pow_service: String, parallelism: usize, backoff_secs: u64) -> Self {
+        let semaphore = Arc::new(Semaphore::new(parallelism));
+        let backoff = time::Duration::from_secs(backoff_secs);
```
Comment (poszu) [inline src/pow/service.rs:21]: NIT:
```suggestion
    pub fn new(k2pow_service: String, parallelism: usize, backoff: time::Duration -> Self {
        let semaphore = Arc::new(Semaphore::new(parallelism));
```

**Review by poszu — COMMENTED** [review]
Generally, looking good! Only few small things left :)

File: service/src/service.rs [hunk 15]
```diff
@@ -110,18 +117,21 @@ pub struct PostService {
     threads: post::config::Cores,
     pow_flags: RandomXFlag,
     proof_generation: Mutex<ProofGenProcess>,
+    remote_k2pow_config: Option<K2powConfig>,
 
     stop: Arc<AtomicBool>,
 }
 
 impl PostService {
+    #[allow(clippy::too_many_arguments)]
```
Comment (poszu) [inline service/src/service.rs:126]: I think it's not needed anymore :)
```suggestion
```

**Review by poszu — APPROVED** [review]
There is one small thing still open: https://github.com/spacemeshos/post-rs/pull/362#discussion_r1782241904.

Otherwise LGTM:)
